### ISSUE TITLE
Steam Deck: Sleeping and then Waking Stops Audio

### ISSUE THREAD

**Issue by IBNobody** [hrydgard/ppsspp#15508]
### Game or games this happens in

ULJM-05781 Valkiria Chronicles 3 Extra Edition (Patched)

### What area of the game / PPSSPP

When putting the Steam Deck to sleep and waking it up, ppsspp no longer plays audio.

### What should happen

When the SD goes to sleep and wakes up, ppsspp does not resume playing audio. 

Steps to reproduce:

1. Launch a game.
3. Put the Steam Deck to sleep.
4. Wake the Steam Deck.

It looks like the SD creates a new audio device, or ppsspp detects the audio device after a sleep state as being new.

If I change the `Device` setting from `Auto` to the selected output and put the system to sleep, the setting will reset back to `Auto`. If I go to set the `Device` again, it shows back up as `Auto`, and the selection list shows new entries incremented numerically with (#). Selecting that entry resumes audio.

Before Sleep and Wake:

![20220428121742_1](https://user-images.githubusercontent.com/5863888/165818889-55e6cda1-52d3-4a29-bb27-b9dab9cd5723.jpg)

After Sleep and Wake:

![20220428123449_1](https://user-images.githubusercontent.com/5863888/165818887-39d4be6b-61f8-4c97-b18f-b8734d402e0a.jpg)

### Logs

Not sure where ppsspp stores flatpak logs on the SD...

### Platform

Linux / BSD

### Mobile phone model or graphics card

Steam Deck

### PPSSPP version affected

v1.12.3 (Flatpak)

### Graphics backend (3D API)

Vulkan

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [ ] Try resetting settings or older versions and include if the issue is related.
- [X] Try without any cheats and without loading any save states.
- [X] Include logs or screenshots of issue.

**Comment by IBNobody** [comment]
I did some more testing by building the latest master repo into a Flatpak, and I am still experiencing this. I also noted that I hear a little bit of audio immediately when the Steam Deck turns back on, but then it immediately dies.

Also, running the latest Win64 binary with Proton Experimental also corrects the issue. My guess is that Proton/Wine is handling the audio drivers a little better.

**Comment by hrydgard** [comment]
Interesting. Maybe we are not handling some error from SDL properly, or similar...

**Comment by IBNobody** [comment]
I enabled logging, set them to notice, and opened up the logs from the dev menu. I didn't see anything standing out to me. Is there another log I should be looking for?

**Comment by IBNobody** [comment]
Found something. Ran the flatpak from the commandline and saw the debug info. This line pops up every time I wake the Steam Deck.

`24:19:856 idle0        I[AUDIO]: SDL/SDLMain.cpp:117 SDL: Trying a different audio device`

So my guess is that it doesn't try long enough after a system wake, or it finds the wrong audio device.

**Comment by IBNobody** [comment]
No. I was wrong, and I have made a breakthrough. The issue is being caused by the fact that the Steam Deck has two separate audio stream devices: speakers and headphones. When the SD goes to sleep, PPSSPP sees an audio device change and automatically switches to the new audio device. However, it must see the headphone device first and switches the audio output to the headphones.

**Comment by hrydgard** [comment]
Oh interesting. I would have hoped that SDL did the right thing internally but sounds like we have to help it do the right thing somehow.

**Comment by IBNobody** [comment]
> Oh interesting. I would have hoped that SDL did the right thing internally but sounds like we have to help it do the right thing somehow.

Yeah. Let me do some more experimentation today on the behavior and get you a full rundown of what happens when I put the SD to sleep with headphones, speakers, or a BT audio device active, with audio device names as reported my PPSSPP.

**Comment by IBNobody** [comment]
@hrydgard I have a work around. Disabling **"Use new audio device"** and keeping **"Device"** set to `Auto` fixes things. PPSSPP will correctly play audio out of the speakers or headphones, and will continue to play audio out of the speaker or headphones after the SD sleeps and wakes up. Bonus, it correctly switches between speaker and headphones if I plug or unplug my physical headphones in.

With I think what is happening is that with **"Use new audio device"** enabled, PPSSPP gets confused.

1. PPSSPP Audio device is set to Auto, and uses `Raven/Raven2/FireFlight/Renior Audio Processor Speaker` or `Raven/Raven2/FireFlight/Renior Audio Processor Headphones` depending on if headphones are physically plugged into the SD.
2. The SD goes to sleep and wakes up.
3. PPSSPP detects that new audio devices are present and names them `Raven/Raven2/FireFlight/Renior Audio Processor Headphones (2)` and `Raven/Raven2/FireFlight/Renior Audio Processor Speakers (2)`
4. Since **"Use new audio device"** is set to `True`, PPSSPP thinks it needs to switch to these new audio devices.
5. PPSSPP switches to using `Raven/Raven2/FireFlight/Renior Audio Processor Headphones (2)` because that is the first new device it sees.
6. The end result is that I don't have audio coming out my speakers. :(

I think the permanent solution would be to add some filtering in on the **"Use new audio device"** subsystem. Rather than allowing it to add new devices with a number (`Raven/Raven2/FireFlight/Renior Audio Processor Headphones (2)`), it should filter that number (` (2)`) out of the name and not switch to it automatically.

**Comment by hrydgard** [comment]
Nice work figuring that out!

It's got to be SDL doing the renaming though, annoying... You think we could just detect the steam deck and disable "Use new audio device" by default, would that be good enough?

**Comment by IBNobody** [comment]
Let me find my BT headphones and see how those are handled.

I don't know if I would disable the setting automatically for the SD, but one thing that I did was submit a [PR ](https://github.com/dragoonDorise/EmuDeck/pull/214) to one of the emulation scripts for the SD that sets everything up. In that script, I disabled the setting. So people using EmuDeck to set up all the emulators would get this as a default setting.

**Comment by unknownbrackets** [comment]
I think what's happening is this, from PPSSPP's perspective:

 * You put the device to sleep.
 * SDL tells PPSSPP that the speaker has been unplugged (as if it were USB.)
 * SDL tells PPSSPP that the headphone jack has been unplugged (I guess it's a dongle in its mind.)
 * At this point, PPSSPP is no mouth or nose, it cannot make any noise.
 * You wake the device.
 * SDL tells PPSSPP a new device has been added.  As if it's a new speaker, it's named "Speaker (4)".
 * SDL tells PPSSPP a new device has been added.  Listed as a new device, "Headphone (4)".

It could be SDL or even your system that's naming it using that number.  If my assumption is correct, the headphone notification is always (or worse, usually?) the last one, so it's the "newest" new device, and PPSSPP switches to it.  From PPSSPP's perspective, it can't really tell this situation apart from a USB dongle being plugged in that's never been plugged in before.

It may also be dangerous to ignore the parenthesis in all cases.  I use an ultrawide monitor in a PBP configuration (i.e. as if it's two separate hardware monitors.)  Each one comes through with a separate audio device (audio over HDMI/DisplayPort) and both audio devices have the same base name.  I'd guess that SDL or whatever is adding the "(2)" and so forth is doing it for people in my situation, who have two devices which can be concurrently connected with the same name.  My monitor has a headphone jack and speakers, but it only routes through one active HDMI connection at a time.

That said, it would probably be sensible to reselect a device if:
 * No device exists with the name of the previously selected device.
 * The new device candidate has the same name as the current, if you strip off ` (##)` from the end.

That might, at least, allow someone to force PPSSPP to use the speakers or the headset.

To make "Use new audio device" usable in this situation, we could remember previously plugged devices and re-select a same named device if you went down to zero and then multiple new devices appeared within a certain time interval (I assume both are "appearing" as new devices within the same few hundred milliseconds.)  There's probably lots of ways that could go wrong for people wanting actual new devices to be recognized, though...

-[Unknown]

**Comment by IBNobody** [comment]
So, not quite sure what scenario **"Use new audio device"** would be used for on the SD. **"Device"** set to `auto` seems to let the OS handle the audio selection, and this seems to work 100%.

I added my BT speaker as a new device, and PPSSPP switched to the BT speaker. I put the SD to sleep and woke it up, and PPSSPP began playing audio through my SD speaker until the SD could connect to the BT speaker. Once the connection was made, PPSSPP began playing through the BT speaker. In the SD settings, I have the ability to change what audio device to use, and PPSSPP's audio plays out of whatever output I tell the SD to use.

So again, I don't know what **"Use new audio device"** would be used for here. Is this setting more for Windows or Android rather than Linux?

**Comment by unknownbrackets** [comment]
Well, it was added for #11727.  All the variants of Linux and BSD probably behave differently here for what the "default" is that Auto uses, but it sounds like the Steam Deck does what one would want.

On Windows (doesn't use SDL), where it's more of a way to not change to a new default device automatically.  That's just because the WASAPI API specifically tells us when the auto device changes.  It's not available on Android at all.

Sounds like disabling this on the Steam Deck would make sense.

-[Unknown]

**Comment by IBNobody** [comment]
> Sounds like disabling this on the Steam Deck would make sense.
> 
> -[Unknown]

And I would specifically only disable this on the SD running its SteamOS 3.0 / Arch as I have not tested Win10 / Win11 on it.
